### ISSUE TITLE
Requests Timed out

### ISSUE THREAD

**Issue by FrancescoSaverioZuppichini** [qdrant/qdrant-client#394]
Hi there, using the client like this

```python
from qdrant_client import QdrantClient

url = "https://<QDRANT_SERVER>"
client = QdrantClient(
    url=url,
    timeout=10,
)

client.<do_stuff>

```

results in timeout in every requests, if I use my local one running I don't have any issue.

If I use `curl` I don't have timeouts, I was wondering if you have any idea why calling the APIs with the sdks doesn't work but with `curl` yes. Maybe some options for https in the client?

Thanks you so much.

I cannot share the server URL because is deployed on Azure in our private network, but it is just qdrant in docker

Checking inside the container, there is not log for the requests, while if I use `curl` I can see that it hits the container

**Comment by FrancescoSaverioZuppichini** [comment]
I can confirm it is an issue with the sdks, calling the server directly works. For example, 

```python
import requests
import json

url = f"{url}/collections/test_2"

payload = json.dumps({"vectors": {"size": 100, "distance": "Dot"}})
headers = {"Content-Type": "application/json"}

response = requests.request("PUT", url, headers=headers, data=payload)

print(response.text)

```

works

**Comment by FrancescoSaverioZuppichini** [comment]
looks like one has to set `https = True` even if the `url` has `https` - could be considered a bug?

**Comment by Thiru-GVT** [comment]
Same error here, HTTP works but the client is broken

**Comment by joein** [comment]
Hi

I'll check it, thanks

In the mean time, @Thiru-GVT does setting `https=True` fixes the issue for you as it does for @FrancescoSaverioZuppichini ?

**Comment by joein** [comment]
@FrancescoSaverioZuppichini @Thiru-GVT could you please provide qdrant version you are using and qdrant-client version?

Btw, in some setups azure deployment requires setting port as 443 instead of 6333

**Comment by joein** [comment]
I've checked various configurations on `qdrant-client==1.7.0`

```
client = QdrantClient(url='https://aws.aws.com:6333', api_key='api_key')
client = QdrantClient(url='https://aws.aws.com:6333', api_key='api_key', https=True)
client = QdrantClient(url='https://aws.aws.com:6333')
client = QdrantClient(url='https://aws.aws.com:6333', https=True)

client = QdrantClient(url='https://aws.aws.com:6333', api_key='api_key', prefer_grpc=True)
client = QdrantClient(url='https://aws.aws.com:6333', api_key='api_key', https=True, prefer_grpc=True)
client = QdrantClient(url='https://aws.aws.com:6333', prefer_grpc=True)
client = QdrantClient(url='https://aws.aws.com:6333', https=True, prefer_grpc=True)
```

All of them seems to work correctly and https is enabled if the scheme contains 'https' regardless of `https` flag being set or not

**Comment by scvsh** [comment]
The same problem, `get_collections` works well but `create_collection` times out. Sync and async clients in python

**Comment by Thiru-GVT** [comment]
Im using `1.6.9` but upgraded to `1.7.0` and i still experience the same issue.

I'm using http://localhost:6333/ so not https. Its with the upsert operation

**Comment by FrancescoSaverioZuppichini** [comment]
> @FrancescoSaverioZuppichini @Thiru-GVT could you please provide qdrant version you are using and qdrant-client version?
> 
> Btw, in some setups azure deployment requires setting port as 443 instead of 6333

yes but my assumption was that if I pass a url with protocol `https` then the sdk should understand that I want to use `https` 😓

**Comment by joein** [comment]
@Thiru-GVT could you provide a python code sample to reproduce or just show us the code you are using?

**Comment by joein** [comment]
@FrancescoSaverioZuppichini sorry, I am not sure whether we are on the same track right now

> looks like one has to set https = True even if the url has https - could be considered a bug?

I was not able to reproduce this one.

Could you provide an example which does not work with `<url containing https scheme>`, but works with `<url containing https scheme> + https=True`

If you have some sensitive url, could you please replace sensitive parts with some gibberish so I could test the format of the url?

like if you have <https://fc055abe-cbcf-4b57-a178-224713c9255d.europe-west3-0.gcp.cloud.qdrant.io>
send <https://<uuid>.<words with digits and hyphens>.gcp.cloud.qdrant.io:6333> or something like this ?

**Comment by HaoES** [comment]
@joein I am experiencing the same issue, here is my python code: 

`client = QdrantClient(
    url="https://c445f1e5-8686-4081-8cab-6ef96ceac615.europe-west3-0.gcp.cloud.qdrant.io:6333",
    api_key=os.environ["QDRANT_API"],
    https=True,
)


embeddings = HuggingFaceEmbeddings(model_name="NeuML/pubmedbert-base-embeddings")

vectorstore = Qdrant(
    client=client,
    collection_name="med_assistant",
    embeddings=embeddings,
)
`

and then later when I do
`vectorstore.add_texts(chunks)`
it give me `ResponseHandlingException: The read operation timed out`

**Comment by joein** [comment]
Hi @HaoES 

It seems like you are using some third party library, not pure python client, aren't you?

Is it langchain or maybe something else?

Could you please provide a launchable code sample (with all imports included?) and maybe some data you are using

Collection info might be useful as well

**Comment by HaoES** [comment]
Hey @joein I am using langchain and streamlit, when I use only a short pdf it works without any problem, but when I use multiple files (around 40 pages each) it times out. Here is my whole `app.py` code: 
```
import streamlit as st
from pypdf import PdfReader
from langchain.text_splitter import CharacterTextSplitter
from langchain.embeddings import HuggingFaceEmbeddings
from langchain.vectorstores import Qdrant
import qdrant_client
from qdrant_client import QdrantClient
import os

# create a qdrant_client

client = QdrantClient(
    url="https://c445f1e5-8686-4081-8cab-6ef96ceac615.europe-west3-0.gcp.cloud.qdrant.io:6333",
    api_key=os.environ["QDRANT_API"],
    https=True,
)


embeddings = HuggingFaceEmbeddings(model_name="NeuML/pubmedbert-base-embeddings")

vectorstore = Qdrant(
    client=client,
    collection_name="med_assistant",
    embeddings=embeddings,
)


def get_pdf_text(docs):
    text = ""
    for doc in docs:
        pdf_reader = PdfReader(doc)
        for page in pdf_reader.pages:
            text += page.extract_text()
    return text


def get_chunks(text):
    text_splitter = CharacterTextSplitter(
        separator="\n", chunk_size=1000, chunk_overlap=200, length_function=len
    )
    chunks = text_splitter.split_text(text)
    return chunks


def main():
    st.set_page_config(page_title="Med Prep", page_icon=":medical_symbol:")
    st.header("Med Prep :medical_symbol:")
    st.text_input("Ask questions about your documents:")

    with st.sidebar:
        st.subheader("Your documents")
        docs = st.file_uploader(
            "Upload your course materials here then click 'Process':",
            accept_multiple_files=True,
            type="pdf",
        )
        if st.button("Process"):
            with st.spinner("Processing files ..."):
                # get pdf files:
                raw_text = get_pdf_text(docs)
                # get chunks of texts
                chunks = get_chunks(raw_text)
                # get vectorstore
                vectorstore.add_texts(chunks)


if __name__ == "__main__":
    main()

```
What do you mean by collection info?

**Comment by Wamy-Dev** [comment]
Having a similar problem, embeddings seem to give this error:

vector_size = len(partial_embeddings[0]), IndexError: list index out of range. 

It seems to be related due to me using http, and only fails when using large requests to qdrant for many or large files.

**Comment by romqn1999** [comment]
> Im using `1.6.9` but upgraded to `1.7.0` and i still experience the same issue.
> 
> I'm using http://localhost:6333/ so not https. Its with the upsert operation

I'm uncertain whether my issue aligns with yours, but I encountered a situation where upserting a substantial volume of data exceeded the specified timeout of [5 seconds](https://python-client.qdrant.tech/qdrant_client.qdrant_client#:~:text=timeout%20%E2%80%93%20Timeout%20for%20REST%20and%20gRPC%20API%20requests.%20Default%3A%205.0%20seconds%20for%20REST%20and%20unlimited%20for%20gRPC). To address this, I opted for a solution utilizing gRPC, with the `prefer_grpc=True` setting.

**Comment by joein** [comment]
hi @Wamy-Dev you have an index error, which means that `partial_embeddings` are empty, not sure how it relates to http

**Comment by joein** [comment]
hi @HaoES 

you might need to check [this page](https://qdrant.tech/documentation/faq/database-optimization/#my-requests-are-very-slow-or-time-out-what-should-i-do)

it might be related to network issues / non-optimal configuration of the collection / etc

you can also try to decrease batch size

sharing `collection info` can also be helpful (by this I mean e.g. output of `QdrantClient.get_collection` method)

**Comment by joein** [comment]
hi @FrancescoSaverioZuppichini 

sorry for the late response



> I've checked various configurations on `qdrant-client==1.7.0`
> 
> ```
> client = QdrantClient(url='https://aws.aws.com:6333', api_key='api_key')
> client = QdrantClient(url='https://aws.aws.com:6333', api_key='api_key', https=True)
> client = QdrantClient(url='https://aws.aws.com:6333')
> client = QdrantClient(url='https://aws.aws.com:6333', https=True)
> 
> client = QdrantClient(url='https://aws.aws.com:6333', api_key='api_key', prefer_grpc=True)
> client = QdrantClient(url='https://aws.aws.com:6333', api_key='api_key', https=True, prefer_grpc=True)
> client = QdrantClient(url='https://aws.aws.com:6333', prefer_grpc=True)
> client = QdrantClient(url='https://aws.aws.com:6333', https=True, prefer_grpc=True)
> ```
> 
> All of them seem to work correctly and https is enabled if the scheme contains 'https' regardless of `https` flag being set or not

As I wrote here, I had tried these configurations and they were correct: all of them set `QdrantClient._client._https` flag correctly and constructed `QdrantClient._client.rest_uri`.

If you can't share even an obfuscated uri, then it might be helpful if you could check values of the mentioned variables on your own and write the result here

**Comment by FrancescoSaverioZuppichini** [comment]
sorry for the late reply as well, noticed the problem once again in production - trying to pin the version and maybe increase the timeout

**Comment by joein** [comment]
@FrancescoSaverioZuppichini it seems for me that we can proceed like this infinitely, I suggest you to concat me in discord so we could solve the issue in real-time, and then publish a post-mortem here

To find me in discord go to our [discord channel](https://qdrant.to/discord) and look for `george.panchuk`.

**Comment by FrancescoSaverioZuppichini** [comment]
> @FrancescoSaverioZuppichini it seems for me that we can proceed like this infinitely, I suggest you to concat me in discord so we could solve the issue in real-time, and then publish a post-mortem here
> 
> To find me in discord go to our [discord channel](https://qdrant.to/discord) and look for `george.panchuk`.

thank you so much for the support, apparently colleagues changed the qdrant port without telling me - no s*hit I was getting timeout haha

I'll close it for now since it seems to be fixed

**Comment by Porkechebure** [comment]
Fixed nothing.

This is still alive and causing troubles.

Running a collection indexing and saving vectors with llamaindex and getting this stuff.
Fix it.

> 
> 2024-03-03 22:26:23 +0000   24957 execution          WARNING  [Node_002_qdrant_txt_read in line 0 (index starts from 0)] stderrGenerating embeddings:  45%|####5     | 10/22 [00:02<00:03,  3.62it/s]
> 2024-03-03 22:26:23 +0000   24957 execution          ERROR    Node test in line 0 failed. Exception: Execution failure in 'test': (ResponseHandlingException) timed out.
> Traceback (most recent call last):
>   File "/home/vscode/.local/lib/python3.11/site-packages/httpx/_transports/default.py", line 69, in map_httpcore_exceptions
>     yield
>   File "/home/vscode/.local/lib/python3.11/site-packages/httpx/_transports/default.py", line 233, in handle_request
>     resp = self._pool.handle_request(req)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/vscode/.local/lib/python3.11/site-packages/httpcore/_sync/connection_pool.py", line 216, in handle_request
>     raise exc from None
>   File "/home/vscode/.local/lib/python3.11/site-packages/httpcore/_sync/connection_pool.py", line 196, in handle_request
>     response = connection.handle_request(
>                ^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/vscode/.local/lib/python3.11/site-packages/httpcore/_sync/connection.py", line 101, in handle_request
>     return self._connection.handle_request(request)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/vscode/.local/lib/python3.11/site-packages/httpcore/_sync/http11.py", line 143, in handle_request
>     raise exc
>   File "/home/vscode/.local/lib/python3.11/site-packages/httpcore/_sync/http11.py", line 113, in handle_request
>     ) = self._receive_response_headers(**kwargs)
>         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/vscode/.local/lib/python3.11/site-packages/httpcore/_sync/http11.py", line 186, in _receive_response_headers
>     event = self._receive_event(timeout=timeout)
>             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/vscode/.local/lib/python3.11/site-packages/httpcore/_sync/http11.py", line 224, in _receive_event
>     data = self._network_stream.read(
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/vscode/.local/lib/python3.11/site-packages/httpcore/_backends/sync.py", line 124, in read
>     with map_exceptions(exc_map):
>   File "/usr/local/lib/python3.11/contextlib.py", line 158, in __exit__
>     self.gen.throw(typ, value, traceback)
>   File "/home/vscode/.local/lib/python3.11/site-packages/httpcore/_exceptions.py", line 14, in map_exceptions
>     raise to_exc(exc) from exc
> httpcore.ReadTimeout: timed out
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/home/vscode/.local/lib/python3.11/site-packages/qdrant_client/http/api_client.py", line 101, in send_inner
>     response = self._client.send(request)
>                ^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/vscode/.local/lib/python3.11/site-packages/httpx/_client.py", line 914, in send
>     response = self._send_handling_auth(
>                ^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/vscode/.local/lib/python3.11/site-packages/httpx/_client.py", line 942, in _send_handling_auth
>     response = self._send_handling_redirects(
>                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/vscode/.local/lib/python3.11/site-packages/httpx/_client.py", line 979, in _send_handling_redirects
>     response = self._send_single_request(request)
>                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/vscode/.local/lib/python3.11/site-packages/httpx/_client.py", line 1015, in _send_single_request
>     response = transport.handle_request(request)
>                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/vscode/.local/lib/python3.11/site-packages/httpx/_transports/default.py", line 232, in handle_request
>     with map_httpcore_exceptions():
>   File "/usr/local/lib/python3.11/contextlib.py", line 158, in __exit__
>     self.gen.throw(typ, value, traceback)
>   File "/home/vscode/.local/lib/python3.11/site-packages/httpx/_transports/default.py", line 86, in map_httpcore_exceptions
>     raise mapped_exc(message) from exc
> httpx.ReadTimeout: timed out
>

**Comment by m-navarro93** [comment]
Hi!

I have an additional solution to try. Looking through the SDK code, if you don't set `port=None` when instantiating a `QdrantClient` object, the string `:6333` gets [added onto the request URL](https://github.com/qdrant/qdrant-client/blob/c63c62e6df9763591622d1921b3dfcc486666481/qdrant_client/http/api_client.py#L73).

This was causing timeout issues for us with our hosted database. So...

`https://my-url.com:6333/collections` -> timeout error
`https://my-url.com/collections` -> Success!

Using the code below allowed me to connect to our hosted database.

```Python
qdrant_api_key = "my-api-key"
qdrant_url = "https://my-url.com"
dev_client = QdrantClient(url=qdrant_url, port=None, api_key=qdrant_api_key)
dev_client.get_collections()
```

Hopefully this helps!

**Comment by TalentumSebasG** [comment]
> Hi!
> 
> I have an additional solution to try. Looking through the SDK code, if you don't set `port=None` when instantiating a `QdrantClient` object, the string `:6333` gets [added onto the request URL](https://github.com/qdrant/qdrant-client/blob/c63c62e6df9763591622d1921b3dfcc486666481/qdrant_client/http/api_client.py#L73).
> 
> This was causing timeout issues for us with our hosted database. So...
> 
> `https://my-url.com:6333/collections` -> timeout error `https://my-url.com/collections` -> Success!
> 
> Using the test code below allowed me to connect to our hosted database.
> 
> ```python
> qdrant_api_key = "my-api-key"
> qdrant_url = "https://my-url.com"
> dev_client = QdrantClient(url=qdrant_url, port=None, api_key=qdrant_api_key)
> dev_client.get_collections()
> ```
> 
> Hopefully this helps!

This is the solution, thank you

**Comment by iremnuy** [comment]
Hello ! 

This issue persist for me but I don't get a timeout when I call get_collection, problem occurs during the search request.

Here is the result of get_collection:

status=<CollectionStatus.GREEN: 'green'> optimizer_status=<OptimizersStatusOneOf.OK: 'ok'> vectors_count=1775036 indexed_vectors_count=1775036 points_count=1775036 segments_count=19 config=CollectionConfig(params=CollectionParams(vectors=VectorParams(size=1024, distance=<Distance.COSINE: 'Cosine'>, hnsw_config=None, quantization_config=None, on_disk=True), shard_number=2, sharding_method=None, replication_factor=1, write_consistency_factor=1, read_fan_out_factor=None, on_disk_payload=True, sparse_vectors=None), hnsw_config=HnswConfig(m=16, ef_construct=100, full_scan_threshold=10000, max_indexing_threads=0, on_disk=False, payload_m=None), optimizer_config=OptimizersConfig(deleted_threshold=0.2, vacuum_min_vector_number=1000, default_segment_number=0, max_segment_size=None, memmap_threshold=None, indexing_threshold=20000, flush_interval_sec=5, max_optimization_threads=None), wal_config=WalConfig(wal_capacity_mb=32, wal_segments_ahead=0), quantization_config=None) payload_schema={}

As can be seen the collection consist of 1.8M vectors but I can do the similarity search within 5 secs in qdrant interface.
Also my call sometimes (like 1/20) does not get a timeout error and I get the scored points.
And also a friend that uses exactly the same config and network most of the time does not get a timeout at all.

Here is the call where I stuck : 

hits = self.client.search(
            collection_name=self.collection_name,
            score_threshold=0.6, #can be used for filtering
            query_vector=query_embedding.tolist()[0],
            with_payload=True,
            limit=10,  # Return 5 closest points
            timeout=20000,
        )

Do you have any ideas ??

**Comment by joein** [comment]
Hi @iremnuy 

You're using `on_disk` option with your vectors, which requires fast disks, what are your disk I/OPS ?
Also to speed up the search you might want to look into the quantization options such as scalar or binary quantization

**Comment by iremnuy** [comment]
Data reside in 50 GB SSD, iostat output is like the following : 

Device            r/s     rkB/s   rrqm/s  %rrqm r_await rareq-sz     w/s     wkB/s   wrqm/s  %wrqm w_await wareq-sz     d/s     dkB/s drqm/s  %drqm d_await dareq-sz     f/s f_await  aqu-sz  %util
xvda             9.94     86.00     0.24   2.33    1.77     8.65    5.91    215.81     5.74  49.26    2.62    36.52    0.00      0.00     0.00   0.00    0.00     0.00    0.00    0.00    0.03   1.67

Therefore total IOPS is around 16.

I will check for quantization options, and let you know if it helps, thank you !

**Comment by joein** [comment]
Are you sure that these are the correct stats?
It is extremely slow, even hdds are faster

**Comment by iremnuy** [comment]
I checked my AWS volumes, I have 4 of them with each  100/100/150/180 IOPS respectively, they still seem to be slow, do you have any suggestion on the threshold of IOPS that will satisfy my needs? (Searching through 1.8M vectors in 2-3 secs in my case), I also checked the read/sec graph, it occasionally peeks at 16K IOPS.

**Comment by iremnuy** [comment]
I solved the problem with scalar quantization, I am able to do similarity search among 1.8m vectors in 0.85 secs on average with a ssd that supports up to 3k IOPS, Thank you.

**Comment by SUMITDHAKAD0** [comment]
First create Empty collection on that add data in batches

Batch_size = 500
no_doc = 0
for i in range(0, len(text_chunks), Batch_size):
    print(i)
    # # Add more documents
    uuids = manager.add_documents(text_chunks[i:i+Batch_size])
    no_doc += len(uuids)

**Comment by brandonbiggs** [comment]
If someone else comes across this issue, I wasn't running into a time out issue, but I wasn't able to connect to my qdrant database using the client for two reasons.
1. Like someone mentioned above, I had to set `port=443` https://github.com/qdrant/qdrant-client/issues/394#issuecomment-2008190756
2. And I was also hosting qdrant under a sub url so I had to add `prefix=` when I was instantiating my `QdrantClient` https://github.com/qdrant/qdrant-client/issues/755

Hope this helps.
